File: frontend/src/lib/content-schema.ts
import activitiesRaw from "@/content/activities.json";
import skillTagsRaw from "@/content/skill-tags.json";
import themesRaw from "@/content/themes.json";

export const skillTags = skillTagsRaw;
export type SkillTag = (typeof skillTags)[number];
export const activityThemes = themesRaw;
export type ActivityTheme = (typeof activityThemes)[number];
export const difficultyTiers = ["easy", "medium", "difficult"] as const;
export type DifficultyTier = (typeof difficultyTiers)[number];
const MIN_PASSAGE_SENTENCES = 8;
const MIN_PASSAGE_PARAGRAPHS = 2;
const MIN_PARAGRAPH_SENTENCES = 2;
const SETUP_CUES = [
  "needed to",
  "had to",
  "problem",
  "challenge",
  "question",
  "goal",
  "plan",
  "clue",
  "worried",
  "noticed",
  "task",
  "decided",
  "wanted",
  "tried",
  "began",
  "hoped",
  "set out",
  "wondered",
] as const;
const OUTCOME_CUES = [
  "by the end",
  "learned",
  "as a result",
  "this showed",
  "this helped",
  "improved",
  "solved",
  "understood",
  "agreed",
  "realized",
  "discovered",
  "finally",
  "after that",
  "from then on",
  "knew",
] as const;

export type QuestionType = "multiple-choice" | "short-response";

export type Question = {
  id: string;
  type: QuestionType;
  prompt: string;
  choices?: string[];
};

export type Activity = {
  id: string;
  title: string;
  theme: ActivityTheme;
  difficulty: DifficultyTier;
  passageType: "literary" | "informational" | "poetry";
  missionLabel: string;
  passageTitle: string;
  passageText: string;
  questions: Question[];
  skillTags: SkillTag[];
};

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function isNonEmptyString(value: unknown): value is string {
  return typeof value === "string" && value.trim().length > 0;
}

function sentenceCount(text: string): number {
  return text
    .replace(/\n/g, " ")
    .trim()
    .split(/(?<=[.!?])\s+/)
    .filter((chunk) => chunk.trim().length > 0).length;
}

function paragraphCount(text: string): number {
  return text
    .trim()
    .split(/\n\s*\n/)
    .map((chunk) => chunk.trim())
    .filter((chunk) => chunk.length > 0).length;
}

function splitParagraphs(text: string): string[] {
  return text
    .trim()
    .split(/\n\s*\n/)
    .map((chunk) => chunk.trim())
    .filter((chunk) => chunk.length > 0);
}

function hasAnyCue(text: string, cues: readonly string[]): boolean {
  const lowered = text.toLowerCase();
  return cues.some((cue) => lowered.includes(cue));
}

function parseQuestion(value: unknown, activityId: string): Question {
  if (!isObject(value)) {
    throw new Error(`Question in activity "${activityId}" must be an object.`);
  }

  const id = value.id;
  const type = value.type;
  const prompt = value.prompt;
  const choices = value.choices;

  if (!isNonEmptyString(id)) {
    throw new Error(`Question in activity "${activityId}" has an invalid id.`);
  }
  if (type !== "multiple-choice" && type !== "short-response") {
    throw new Error(`Question "${id}" in "${activityId}" has an invalid type.`);
  }
  if (!isNonEmptyString(prompt)) {
    throw new Error(`Question "${id}" in "${activityId}" has an empty prompt.`);
  }

  if (type === "multiple-choice") {
    if (!Array.isArray(choices) || choices.length < 2 || !choices.every(isNonEmptyString)) {
      throw new Error(`Multiple-choice question "${id}" in "${activityId}" needs at least two choices.`);
    }
    return { id, type, prompt, choices };
  }

  return { id, type, prompt };
}

function parseActivity(value: unknown): Activity {
  if (!isObject(value)) {
    throw new Error("Activity entry must be an object.");
  }

  const id = value.id;
  const title = value.title;
  const theme = value.theme;
  const difficulty = value.difficulty;
  const passageType = value.passageType;
  const missionLabel = value.missionLabel;
  const passageTitle = value.passageTitle;
  const passageText = value.passageText;
  const questionsRaw = value.questions;
  const tagsRaw = value.skillTags;

  if (!isNonEmptyString(id)) throw new Error("Activity id is required.");
  if (!isNonEmptyString(title)) throw new Error(`Activity "${id}" is missing title.`);
  if (!isNonEmptyString(theme) || !activityThemes.includes(theme as ActivityTheme)) {
    throw new Error(`Activity "${id}" has invalid or unsupported theme.`);
  }
  if (passageType !== "literary" && passageType !== "informational" && passageType !== "poetry") {
    throw new Error(`Activity "${id}" has invalid passageType.`);
  }
  if (difficulty !== undefined && (!isNonEmptyString(difficulty) || !difficultyTiers.includes(difficulty as DifficultyTier))) {
    throw new Error(`Activity "${id}" has invalid difficulty.`);
  }
  if (!isNonEmptyString(missionLabel)) throw new Error(`Activity "${id}" is missing missionLabel.`);
  if (!isNonEmptyString(passageTitle)) throw new Error(`Activity "${id}" is missing passageTitle.`);
  if (!isNonEmptyString(passageText)) throw new Error(`Activity "${id}" is missing passageText.`);
  if (passageType === "poetry") {
    const lines = passageText.split("\n").filter((line) => line.trim().length > 0);
    const stanzas = passageText.split("\n\n").filter((s) => s.trim().length > 0);
    if (lines.length < 8) {
      throw new Error(`Poetry activity "${id}" must include at least 8 lines.`);
    }
    if (stanzas.length < 2) {
      throw new Error(`Poetry activity "${id}" must include at least 2 stanzas.`);
    }
  } else {
    if (sentenceCount(passageText) < MIN_PASSAGE_SENTENCES) {
      throw new Error(`Activity "${id}" must include at least ${MIN_PASSAGE_SENTENCES} sentences in passageText.`);
    }
    if (paragraphCount(passageText) < MIN_PASSAGE_PARAGRAPHS) {
      throw new Error(`Activity "${id}" must include at least ${MIN_PASSAGE_PARAGRAPHS} paragraphs in passageText.`);
    }
    const paragraphs = splitParagraphs(passageText);
    if (sentenceCount(paragraphs[0] ?? "") < MIN_PARAGRAPH_SENTENCES) {
      throw new Error(`Activity "${id}" first paragraph must include at least ${MIN_PARAGRAPH_SENTENCES} sentences.`);
    }
    if (sentenceCount(paragraphs[1] ?? "") < MIN_PARAGRAPH_SENTENCES) {
      throw new Error(`Activity "${id}" second paragraph must include at least ${MIN_PARAGRAPH_SENTENCES} sentences.`);
    }
    if (!hasAnyCue(paragraphs[0] ?? "", SETUP_CUES)) {
      throw new Error(`Activity "${id}" first paragraph must include setup/challenge context cues.`);
    }
    if (!hasAnyCue(passageText, OUTCOME_CUES)) {
      throw new Error(`Activity "${id}" must include outcome/reflection cues in the passage.`);
    }
  }
  if (!Array.isArray(questionsRaw) || questionsRaw.length < 2) {
    throw new Error(`Activity "${id}" must include at least two questions.`);
  }
  if (!Array.isArray(tagsRaw) || tagsRaw.length === 0) {
    throw new Error(`Activity "${id}" must include skill tags.`);
  }

  const questions = questionsRaw.map((question) => parseQuestion(question, id));
  const tags = tagsRaw.filter(isNonEmptyString);
  if (tags.length !== tagsRaw.length) {
    throw new Error(`Activity "${id}" has malformed skill tags.`);
  }

  tags.forEach((tag) => {
    if (!skillTags.includes(tag as SkillTag)) {
      throw new Error(`Activity "${id}" has unsupported skill tag "${tag}".`);
    }
  });

  const resolvedDifficulty = (difficulty as DifficultyTier | undefined) ?? "easy";

  return {
    id,
    title,
    theme: theme as ActivityTheme,
    difficulty: resolvedDifficulty,
    passageType,
    missionLabel,
    passageTitle,
    passageText,
    questions,
    skillTags: tags as SkillTag[],
  };
}

function parseActivities(raw: unknown): Activity[] {
  if (!Array.isArray(raw) || raw.length === 0) {
    throw new Error("Seed activities file must be a non-empty array.");
  }

  const parsed = raw.map(parseActivity);
  const ids = new Set(parsed.map((activity) => activity.id));
  if (ids.size !== parsed.length) {
    throw new Error("Seed activities contain duplicate IDs.");
  }
  const sorted = [...parsed].sort((a, b) => a.id.localeCompare(b.id));
  sorted.forEach((activity, index) => {
    activity.difficulty = difficultyTiers[index % difficultyTiers.length];
  });
  return parsed;
}

export const activities = parseActivities(activitiesRaw);

File: backend/app/content_schema.py
import functools
from pathlib import Path
from typing import Literal, Optional

from pydantic import BaseModel, Field


SkillTag = Literal[
    "reading-comprehension",
    "main-idea",
    "inference",
    "sequence",
    "summary",
    "vocabulary",
    "sentence-quality",
    "short-writing",
]

PassageType = Literal["literary", "informational", "poetry"]
QuestionType = Literal["multiple-choice", "short-response"]
DifficultyTier = Literal["easy", "medium", "difficult"]
ReviewStatus = Literal["reviewed", "draft", "rewrite-required"]

REPO_ROOT = Path(__file__).resolve().parents[2]
BACKEND_CONTENT_DIR = Path(__file__).resolve().parent.parent / "content"
FRONTEND_CONTENT_DIR = REPO_ROOT / "frontend" / "src" / "content"

# backend/content/ is the canonical source. frontend/src/content/ is a
# checked-in mirror synced by scripts/sync-content.sh so the Next.js build
# can import the JSON at compile time. Drift between the two is caught by
# backend/tests/test_content_workflow.py.
CONTENT_DIR = BACKEND_CONTENT_DIR

ACTIVITIES_FILE = CONTENT_DIR / "activities.json"
SKILL_TAGS_FILE = CONTENT_DIR / "skill-tags.json"
THEMES_FILE = CONTENT_DIR / "themes.json"
MANIFEST_FILE = CONTENT_DIR / "MANIFEST.json"
REVIEW_STATUS_FILE = CONTENT_DIR / "review-status.json"
MIN_PASSAGE_SENTENCES = 8
MIN_PASSAGE_PARAGRAPHS = 2
MIN_PARAGRAPH_SENTENCES = 2
SETUP_CUES = (
    "needed to",
    "had to",
    "problem",
    "challenge",
    "question",
    "goal",
    "plan",
    "clue",
    "worried",
    "noticed",
    "task",
    "decided",
    "wanted",
    "tried",
    "began",
    "hoped",
    "set out",
    "wondered",
)
OUTCOME_CUES = (
    "by the end",
    "learned",
    "as a result",
    "this showed",
    "this helped",
    "improved",
    "solved",
    "understood",
    "agreed",
    "realized",
    "discovered",
    "finally",
    "after that",
    "from then on",
    "knew",
)


class QuestionModel(BaseModel):
    id: str = Field(min_length=1)
    type: QuestionType
    prompt: str = Field(min_length=1)
    choices: Optional[list[str]] = None
    correctChoice: Optional[str] = None
    skillTag: Optional[SkillTag] = None
    writingSkillTags: list[SkillTag] = Field(default_factory=list)
    answerExplanation: Optional[str] = None
    responseGuidance: Optional[str] = None


class ActivityModel(BaseModel):
    id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    theme: str = Field(min_length=1)
    difficulty: Optional[DifficultyTier] = None
    passageType: PassageType
    missionLabel: str = Field(min_length=1)
    passageTitle: str = Field(min_length=1)
    passageText: str = Field(min_length=1)
    questions: list[QuestionModel] = Field(min_length=2)
    skillTags: list[SkillTag] = Field(min_length=1)
    sourceUrls: list[str] = Field(default_factory=list)


class DeterministicWritingRubricModel(BaseModel):
    completion: str
    relevance: str
    sentence_completeness: str
    skill_specific_checks: list[str]


def _hash_file(path: Path) -> str:
    import hashlib

    return hashlib.sha256(path.read_bytes()).hexdigest()


def load_content_manifest() -> dict:
    """Read MANIFEST.json. Raises ``ValueError`` if missing or malformed."""
    import json

    if not MANIFEST_FILE.exists():
        raise ValueError(f"Content manifest missing at {MANIFEST_FILE}.")
    data = json.loads(MANIFEST_FILE.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError("MANIFEST.json must be a JSON object.")
    if "content_version" not in data or "files" not in data:
        raise ValueError("MANIFEST.json must include content_version and files.")
    if not isinstance(data["files"], dict) or not data["files"]:
        raise ValueError("MANIFEST.json files entry must be a non-empty object.")
    return data


def verify_content_manifest() -> dict:
    """Return the manifest, raising ``ValueError`` if any file's checksum does
    not match the recorded value or any expected file is missing."""
    manifest = load_content_manifest()
    for name, expected in manifest["files"].items():
        file_path = CONTENT_DIR / name
        if not file_path.exists():
            raise ValueError(f"Manifest lists {name} but file is missing.")
        actual = _hash_file(file_path)
        if actual != expected:
            raise ValueError(
                f"Manifest checksum mismatch for {name}: "
                f"expected {expected[:12]}…, got {actual[:12]}…. "
                "Regenerate via python3 -m backend.app.content_cli manifest."
            )
    return manifest


def _count_sentences(text: str) -> int:
    import re

    return len([part for part in re.split(r"(?<=[.!?])\s+", text.replace("\n", " ").strip()) if part])


def _count_paragraphs(text: str) -> int:
    import re

    return len([part for part in re.split(r"\n\s*\n", text.strip()) if part.strip()])


def _split_paragraphs(text: str) -> list[str]:
    import re

    return [part.strip() for part in re.split(r"\n\s*\n", text.strip()) if part.strip()]


def _count_sentences_in_block(text: str) -> int:
    return _count_sentences(text)


def _has_any_cue(text: str, cues: tuple[str, ...]) -> bool:
    lowered = text.lower()
    return any(cue in lowered for cue in cues)


def _apply_default_difficulties(activities: list[ActivityModel]) -> None:
    tiers: tuple[DifficultyTier, DifficultyTier, DifficultyTier] = ("easy", "medium", "difficult")
    sorted_activities = sorted(activities, key=lambda item: item.id)
    for index, activity in enumerate(sorted_activities):
        if activity.difficulty is None:
            activity.difficulty = tiers[index % len(tiers)]


def load_review_statuses() -> dict[str, ReviewStatus]:
    """Load the explicit editorial status for every canonical activity."""
    import json

    raw = json.loads(REVIEW_STATUS_FILE.read_text(encoding="utf-8"))
    if not isinstance(raw, dict):
        raise ValueError("review-status.json must be a JSON object.")
    allowed: tuple[ReviewStatus, ...] = ("reviewed", "draft", "rewrite-required")
    statuses: dict[str, ReviewStatus] = {}
    for status in allowed:
        ids = raw.get(status)
        if not isinstance(ids, list):
            raise ValueError(f'review-status.json needs a "{status}" array.')
        for activity_id in ids:
            if not isinstance(activity_id, str) or not activity_id:
                raise ValueError(f'review-status.json "{status}" entries must be non-empty strings.')
            if activity_id in statuses:
                raise ValueError(f'Duplicate review status for activity "{activity_id}".')
            statuses[activity_id] = status
    extra = sorted(set(raw) - set(allowed))
    if extra:
        raise ValueError(f"Unsupported review status groups: {', '.join(extra)}")
    return statuses


def load_seed_activities() -> list[ActivityModel]:
    import json

    allowed_themes = json.loads(THEMES_FILE.read_text(encoding="utf-8"))
    raw = json.loads(ACTIVITIES_FILE.read_text(encoding="utf-8"))
    activities = [ActivityModel.model_validate(item) for item in raw]
    statuses = load_review_statuses()
    reviewed_ids = {activity_id for activity_id, status in statuses.items() if status == "reviewed"}
    for activity in activities:
        if activity.id in reviewed_ids and activity.difficulty is None:
            raise ValueError(f'Reviewed activity "{activity.id}" needs explicit difficulty.')
    _apply_default_difficulties(activities)

    seen_ids: set[str] = set()
    for activity in activities:
        if activity.id in seen_ids:
            raise ValueError(f"Duplicate activity id: {activity.id}")
        seen_ids.add(activity.id)
        if activity.theme not in allowed_themes:
            raise ValueError(f'Activity "{activity.id}" has unsupported theme "{activity.theme}".')
        if activity.passageType == "poetry":
            lines = [line for line in activity.passageText.split("\n") if line.strip()]
            stanzas = [s.strip() for s in activity.passageText.split("\n\n") if s.strip()]
            if len(lines) < 8:
                raise ValueError(f'Poetry activity "{activity.id}" must include at least 8 lines.')
            if len(stanzas) < 2:
                raise ValueError(f'Poetry activity "{activity.id}" must include at least 2 stanzas.')
        else:
            if _count_sentences(activity.passageText) < MIN_PASSAGE_SENTENCES:
                raise ValueError(
                    f'Activity "{activity.id}" must include at least {MIN_PASSAGE_SENTENCES} sentences in passageText.'
                )
            if _count_paragraphs(activity.passageText) < MIN_PASSAGE_PARAGRAPHS:
                raise ValueError(
                    f'Activity "{activity.id}" must include at least {MIN_PASSAGE_PARAGRAPHS} paragraphs in passageText.'
                )
            paragraphs = _split_paragraphs(activity.passageText)
            if _count_sentences_in_block(paragraphs[0]) < MIN_PARAGRAPH_SENTENCES:
                raise ValueError(
                    f'Activity "{activity.id}" first paragraph must include at least {MIN_PARAGRAPH_SENTENCES} sentences.'
                )
            if _count_sentences_in_block(paragraphs[1]) < MIN_PARAGRAPH_SENTENCES:
                raise ValueError(
                    f'Activity "{activity.id}" second paragraph must include at least {MIN_PARAGRAPH_SENTENCES} sentences.'
                )
            if not _has_any_cue(paragraphs[0], SETUP_CUES):
                raise ValueError(f'Activity "{activity.id}" first paragraph must include setup/challenge context cues.')
            if not _has_any_cue(activity.passageText, OUTCOME_CUES):
                raise ValueError(f'Activity "{activity.id}" must include outcome/reflection cues in the passage.')

        for question in activity.questions:
            if question.skillTag is not None and question.skillTag not in activity.skillTags:
                raise ValueError(
                    f'Question "{question.id}" in activity "{activity.id}" uses skill tag '
                    f'"{question.skillTag}" that is not listed on the activity.'
                )
            if question.type == "multiple-choice":
                if question.choices is None or len(question.choices) < 2:
                    raise ValueError(
                        f'Multiple-choice question "{question.id}" in activity "{activity.id}" needs at least two choices.'
                    )
                if not question.correctChoice or question.correctChoice not in question.choices:
                    raise ValueError(
                        f'Multiple-choice question "{question.id}" in activity "{activity.id}" needs a valid correctChoice.'
                    )
            else:
                if question.choices:
                    raise ValueError(
                        f'Short-response question "{question.id}" in activity "{activity.id}" should not define choices.'
                    )
                if question.correctChoice:
                    raise ValueError(
                        f'Short-response question "{question.id}" in activity "{activity.id}" should not define correctChoice.'
                    )
            if question.type != "short-response" and question.writingSkillTags:
                raise ValueError(
                    f'Question "{question.id}" in activity "{activity.id}" may only define '
                    "writingSkillTags for a short response."
                )

    return activities


@functools.lru_cache(maxsize=1)
def list_seed_activities() -> tuple[ActivityModel, ...]:
    return tuple(load_seed_activities())


def list_seed_themes() -> list[str]:
    import json

    raw = json.loads(THEMES_FILE.read_text(encoding="utf-8"))
    if not isinstance(raw, list) or not raw:
        raise ValueError("themes.json must be a non-empty array.")
    parsed = [str(item) for item in raw if isinstance(item, str) and item.strip()]
    if len(parsed) != len(raw):
        raise ValueError("themes.json must include non-empty strings only.")
    if len(set(parsed)) != len(parsed):
        raise ValueError("themes.json contains duplicate values.")
    return parsed


def list_seed_difficulty_tiers() -> list[str]:
    return ["easy", "medium", "difficult"]


def get_seed_activity(activity_id: str) -> ActivityModel:
    for activity in list_seed_activities():
        if activity.id == activity_id:
            return activity
    raise ValueError(f'Activity "{activity_id}" not found.')
